### PR TITLE
For Mentor Review | cli-files | Farzaneh Haghani

### PR DESCRIPTION
**Pull request by farzaneh-haghani**
## Self checklist

- [x] I have titled my PR with `For Mentor Review | {Project Name} | {My Name}`
- [x] I have included a list of changes
- [x] I have tested my changes
- [x] I have explained my changes
- [x] I have asked someone to review my changes

## Description
Made `go-ls` and `go-cat` by go tools

### Change list
- Getting paths by `flag.Args`
- Checking if they are valid by `os.Stat`
- Checking if they are directory by `IsDir`
- Getting the list of directories by `os.ReadDir`
- Scanning the files through buffer for using less memory

### CODE REVIEW

**Comment by farzaneh-haghani** [comment]
regarding [Make `go-ls -h` include a helpful description] in the ReadMe
Do you mean I have to implement `ls -lh` because `ls -h` is the same as `ls`?
`-h` means human readable size (B, KB...). Do you mean this by description?

File: projects/cli-files/go-cat/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		if err != nil {
+			return fmt.Errorf("%s: no such file or directory", path)
+		}
+
+		isDirectory := isPathValid.IsDir()
+		if isDirectory {
+			return fmt.Errorf("%s: is a directory", path)
+		}
+
+		file, err := os.Open(path)
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: Have a read about [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) bugs - in general we try to do the thing (here: open the file) and then if it went wrong try to interpret the error (by looking at the error and seeing if it's because the file didn't exist, was a directory, etc), rather than checking in advance whether we expect something to work. The reason being: It could be that someone deletes the file, or replaces the file with a directory, or somethings else _between_ when we check and when we open, whereas if we just try to open and are told "it's a directory, I can't open this" we know this was definitely the problem.

File: projects/cli-files/go-cat/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for i := 0; scanner.Scan(); i++ {
+			if pathArr[0] == "-n" {
+				text := fmt.Sprint("     ", i+1, "  ", scanner.Text()+"\n")
+				os.Stdout.Write([]byte(text))
+			} else {
+				os.Stdout.Write([]byte(scanner.Text() + "\n"))
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: What should happen if you `cat` a file which _doesn't_ end in a newline?

File: projects/cli-files/go-cat/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+	var i int
+	if pathArr[0] == "-n" {
+		i = 1
+	} else {
+		i = 0
+	}
+
+	for i < len(pathArr) {
+		path := pathArr[i]
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: I'd generally recommend using a for-each loop rather than keeping a counter - is there a reason you didn't here?

File: projects/cli-files/go-cat/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+func Execute() error {
+	pathArr := os.Args[1:]
+
+	if len(pathArr) == 0 {
+		return fmt.Errorf("did not provide any path")
+	}
+
+	var i int
+	if pathArr[0] == "-n" {
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: This works well when there's just one flag (and when we expect flags to come before arguments) - if you wanted more flags or more complicated behaviour, I'd recommend looking at https://pkg.go.dev/flag or https://github.com/spf13/cobra or https://github.com/jessevdk/go-flags but this works fine for what you're doing here :)

File: projects/cli-files/go-cat/cmd/root.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+func Execute() error {
+	pathArr := os.Args[1:]
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: This is an interesting naming - while it's an array of _mostly_ paths, it may also contain flags. I'd maybe think about calling this `args` and then making a new slice called `pathArr` which is either the same as, or a subset of `args`, rather than calling this `pathArr` even though it sometimes contains things that aren't paths.

File: projects/cli-files/go-ls/cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		directoryOutput(path, comma)
+		return nil
+	}
+
+	var i int
+	if comma {
+		i = 1
+	} else {
+		i = 0
+	}
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: This works here for the "there's exactly one flag" case, but feels a little hard to follow and complex - give some thought to how you could structure this to really worry about each case once. Something like:

```go
if len(pathArr) > 0 && pathArr[0] == "-m" {
    comma = true
    pathArr = pathArr[1:]
}
if len(pathArr) == 0 {
    pathArr = []string{"."}
}
```

The core idea here is to try to make each step of the problem reduce down to a similar problem if we can. "Stripping a leading -m" is one standalone step and after we do that the problem is just normalise a maybe-empty slice to have a default value.

If our goal is to end up with a slice of path strings, and a boolean saying whether to output commas, we can do the looking for `-m`s (removing them and setting `comma` to `true`) and then treat the rest of the arguments as either paths or "no path present". Now we just need to worry about two special cases: If there's a `-m`, and if there are no paths.

This feels simpler to reason about than your current solution which treats no path, -m, no paths and -m, or some paths and no -m, and some paths and a -m each as different branches we need to consider.

File: projects/cli-files/go-ls/cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+	for i < len(pathArr) {
+
+		isPathValid, err := os.Stat(pathArr[i])
+		if err != nil {
+			return fmt.Errorf("%s no such file or directory", pathArr[i])
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: Do we know the error was caused by the file/directory not existing? Or is it possible something else went wrong?

In general, unless we're confident we understand what went wrong (or we're actively trying to hide information, which is also sometimes valid), it's useful to preserve original errors in what we display rather than replacing them.

File: projects/cli-files/go-ls/cmd/root.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		isDirectory := isPathValid.IsDir()
+		if !isDirectory {
+			if comma && i < len(pathArr)-1 {
+				os.Stdout.Write([]byte(pathArr[i] + ", "))
+			} else {
+				os.Stdout.Write([]byte(pathArr[i] + "  "))
+			}
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: I'd generally aim to factor out what's common here, so probably would write either:
```suggestion
			os.Stdout.Write([]byte(pathArr[i]))
			if comma && i < len(pathArr)-1 {
				os.Stdout.Write(b", ")
			} else {
				os.Stdout.Write(b"  ")
			}
```

or
```suggestion
			var suffix string
			if comma && i < len(pathArr)-1 {
				suffix = ", "
			} else {
				suffix = "  "
			}
			os.Stdout.Write([]byte(pathArr[i] + suffix))
```

This makes it more clear what's the same about both cases (in both cases you're printing the filename), and what's different (the suffix you're printing after).

File: projects/cli-files/go-ls/cmd/root.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		} else {
+			dirPath1 := len(pathArr) > 1 && !comma
+			dirPath2 := len(pathArr) > 2 && comma
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: I'm not sure from their names what `dirPath1` and `dirPath2` are meant to mean - can you think of names that make this more clear?

(Or maybe this goes away if you switch how you populated `pathArr` above, so that rather than needing to worry about whether some of the things in `pathArr` are not paths, you end up with `pathArr` always containing just the list of paths, so not needing to worry about `comma` when doing this check :))

File: projects/cli-files/go-ls/cmd/root.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+			if dirPath1 || dirPath2 {
+				os.Stdout.Write([]byte("\n" + pathArr[i] + ":\n")) // situation: go-ls assets cmd
+			}
+			directoryOutput(pathArr[i], comma)
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: What happens if `directoryOutput` returns a non-nil error? What should happen?

**Review by illicitonion — COMMENTED** [review]
This is generally looking really good, nice job! A lot of the comments I left about go-cat also apply to go-ls, so give some thought to that when you go over things :) Happy to talk through anything that's confusing or not clear!

File: projects/cli-files/go-cat/cmd/root.go (hunk 3 above)
Comment (farzaneh-haghani) [reply projects/cli-files/go-cat/cmd/root.go]: I will change it. There is no reason, just a habit for using for loop or if-else more than for-each or switch-case.

File: projects/cli-files/go-cat/cmd/root.go (hunk 2 above)
Comment (farzaneh-haghani) [reply projects/cli-files/go-cat/cmd/root.go]: If there wasn't a newline, it shows a % at the end. Also if I cat more than one file, the next file will start immediately at the end of first file.

**Comment by farzaneh-haghani** [comment]
@illicitonion I made some changes and would appreciate it if you could review them whenever you have time.

File: projects/cli-files/go-cat/cmd/root.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+	var pathArr []string
+
+	if *numberLine && *help {
+		pathArr = os.Args[3:]
+	} else if !*numberLine && !*help {
+		pathArr = os.Args[1:]
+	} else {
+		pathArr = os.Args[2:]
+	}
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: There's a handy function `flag.Args()` which returns all of the remaining arguments _after_ any flags that were parsed - this may help you simplify your code here a bit :)

File: projects/cli-files/go-cat/cmd/root.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+	if *help {
+		err := OpenFile(helpFilePath, false)
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: This works here, but it relies on there being a file named `assets/help.txt` relative to the working directory wherever this program is run. If you `cd` to `/tmp` and try to `go-cat -h` it will fail, or if you just upload `go-cat` somewhere and someone downloads it, `go-cat -h` will fail. Can you think how to avoid this problem?

File: projects/cli-files/go-ls/cmd/root.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+		err := DirectoryOutput(path, *comma)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	for i, path := range pathArr {
+
+		isPathValid, err := os.Stat(path)
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: `isPathValid` sounds like it would contain a boolean, but it doesn't - can you think of a more clear name for this variable?

File: projects/cli-files/go-ls/cmd/root.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+		isDirectory := isPathValid.IsDir()
+		if !isDirectory {
+			var suffix string
+			if *comma && i < len(pathArr)-1 {
+				suffix = ", "
+			} else {
+				suffix = "  "
+			}
+			os.Stdout.Write([]byte(path + suffix))
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: Nice extracting of the difference in cases into a variable here, this is nice and easy to read an dunderstand!

File: projects/cli-files/go-ls/cmd/root.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+			}
+			err := DirectoryOutput(path, *comma)
+			if err != nil {
+				return err
+			}
+			if i < len(pathArr)-1 {
+				fmt.Print("\n")
+			}
+		}
+		i++
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: What does this `i++` do?

File: projects/cli-files/go-ls/cmd/root.go [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+func DirectoryOutput(path string, comma bool) error {
+
+	list, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(list); i++ {
+		if comma && i < len(list)-1 {
+			os.Stdout.Write([]byte(list[i].Name() + ",\t"))
```
Comment (illicitonion) [inline projects/cli-files/go-ls/cmd/root.go]: Could you do similar to above to make more clear what's the same and what's different in the different cases?

**Review by illicitonion — COMMENTED** [review]
This is looking really good, congrats!

Next up I'd recommend trying to write some tests for these two programs :)

File: projects/cli-files/go-cat/cmd/root.go [hunk 17]
```diff
@@ -2,34 +2,29 @@ package cmd
...
 import (
 	"bufio"
+	"embed"
 	"flag"
 	"fmt"
 	"os"
 )
 
+//go:embed help.txt
+var helpFile embed.FS
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: Ooh, interesting use of embed, this is cool!

One of the interesting things about embed is you can embed as different types - if you declared this variable to be a `string` instead of an `embed.FS`, you could just use it as a string rather than needing to call `.ReadFile` and handle a potential error.

**Review by illicitonion — COMMENTED** [review]
This is looking great! Time for some test? :)

File: projects/cli-files/go-cat/cmd/root.go [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+	fileNames, numberLine, help := ParsingFlags()
+
+	err := flagHandler(os.Stdout, fileNames, numberLine, help)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ParsingFlags() (fileNames []string, numberLine *bool, help *bool) {
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go:25]: I would expect this to return non-pointers, i.e. a `[]string, bool, bool` rather than `[]string, *bool, *bool` - generally returning a pointer happens either because what you're returning is optional (so may be `nil`), or is a value shared with other pieces of code (so multiple references to exactly the same value need to exist).

File: projects/cli-files/go-cat/cmd/root.go [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+	for _, path := range fileNames {
+		err := OpenFile(w, path, *numberLine)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func OpenFile(w io.Writer, path string, numberLine bool) error {
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root.go]: This function feels misleadingly named - it doesn't just _open_ a file, it also _prints_ the file?

File: projects/cli-files/go-cat/cmd/root_test.go [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+		{"true","false","Read the help file with numbers and without h flag",numberLine,help,[]string{"help.txt"},"\t1       The options are as follows:\n     2       \n     3       -b      Number the non-blank output lines, starting at 1.\n     4       \n     5       -e      Display non-printing characters (see the -v option), and display a\n     6               dollar sign (‘$’) at the end of each line.\n     7       \n     8       -l      Set an exclusive advisory lock on the standard output file\n     9               descriptor.  This lock is set using fcntl(2) with the F_SETLKW\n     10               command.  If the output file is already locked, cat will block\n     11               until the lock is acquired.\n"},
+	}
+
+	t.Run("Reading the files", func(t *testing.T) {
+		for _, test := range flagTests {
+			flag.Set("n", test.flagN)
+			flag.Set("h", test.flagH)
+			flag.Parse()
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root_test.go]: This works, but I would expect you to just pass booleans to `flagHandler` rather than going through a whole string -> flag -> flag value loop. `flagHandler` doesn't really care _where_ those booleans came from, just that they were set, so using flags here are more complicated than this test needs to be.

If you get rid of the flag stuff, and put names and one-per-line, your test-cases will end up being things like:

```go
{
    numberLines: true,
    help: false,
    fileNames: []string{"something"},
    wantOutput: "1\tThis is exciting output",
},
{
    numberLines: false,
    help: false,
    fileNames: []string{"something"},
    wantOutput: "This is exciting output",
},
```

which is generally pretty easy to see what's going on.

File: projects/cli-files/go-cat/cmd/root_test.go [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+	flagTests := []struct {
+		flagN      string
+		flagH      string
+		name       string
+		numberLine *bool
+		help       *bool
+		fileNames  []string
+		expect     string
+	}{
+		{"false", "true", "Read the help file by using h flag", numberLine, help, []string{}, "\n"},
```
Comment (illicitonion) [inline projects/cli-files/go-cat/cmd/root_test.go]: I'd typically expect each of the values in this struct to be on its own line, and for each one to have the field name before the value, to make it easier to see what they all do at a glance.

(I can't at a glance work out what test cases you are and aren't covering here :))

**Review by illicitonion — COMMENTED** [review]
Good progress! All of my comments apply to both projects :)

**Comment by farzaneh-haghani** [comment]
> Good progress! All of my comments apply to both projects :)

@illicitonion Thank you for the feedback 👍 
 I applied them.
